### PR TITLE
Metric Backend and Huggingface evaluator support

### PR DESCRIPTION
**Pull request by trajepl**
## Describe your changes

This PR introduce the concept of metric backend where we define the interface to make some metrics backend like huggingface/scipy and etc are able to be plugged and played. 

In this PR, we only implement huggingface backend which support many metrics from community including classification and generative tasks. 

## Checklist before requesting a review
- [x] Add unit tests for this change.
- [x] Make sure all tests can pass.
- [x] Update documents if necessary.
- [x] Format your code by running `pre-commit run --all-files`

### CODE REVIEW

File: olive/evaluator/metric.py [hunk 1]
```diff
@@ -79,6 +79,7 @@ def validate_goal(cls, v, values):
 class Metric(ConfigBase):
     name: str
     type: MetricType
+    backend: Optional[str] = "torch_metrics"
```
Comment (guotuofeng) [inline olive/evaluator/metric.py:82]: >Optional [](http://example.com/codeflow?start=13&length=8)

could the backend be None?

Comment (trajepl) [reply olive/evaluator/metric.py:82]: Yes, for customized metric, it should be None.

File: olive/evaluator/metric.py [hunk 2]
```diff
@@ -91,6 +92,20 @@ def get_sub_type_info(self, info_name, no_priority_filter=True, callback=lambda
...
+        if values["type"] == MetricType.CUSTOM:
+            return None
+        from olive.evaluator.metric_backend import MetricBackend
+
+        assert v in MetricBackend.registry, f"Backend {v} is not in {list(MetricBackend.registry.keys())}"
+        if v == "huggingface_metrics":
+            try:
+                import evaluate  # noqa: F401
+            except ImportError:
+                raise ImportError("Please install the huggingface/evaluate package to use huggingface metrics.")
```
Comment (guotuofeng) [inline olive/evaluator/metric.py:106]: it seems there are duplicated code in backend py file, could we reuse it?

File: olive/evaluator/metric_backend.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+    @staticmethod
+    def _default_config() -> Dict[str, ConfigParam]:
+        return {}
+
+    @abstractmethod
+    def measure(self, preds, targets, sub_metric) -> SubMetricResult:
+        raise NotImplementedError()
+
+    def combine_measure(self, preds, targets, metrics) -> MetricResult:
```
Comment (guotuofeng) [inline olive/evaluator/metric_backend.py:28]: >metrics [](http://example.com/codeflow?start=46&length=7)

add type notation for metrics and sub_metric?

Comment (guotuofeng) [inline olive/evaluator/metric_backend.py:28]: >combine_measure [](http://example.com/codeflow?start=8&length=15)

the name seems a little bit confusing.

Comment (trajepl) [reply olive/evaluator/metric_backend.py]: Updated.

Comment (trajepl) [reply olive/evaluator/metric_backend.py]: Added.

File: olive/evaluator/metric.py (hunk 2 above)
Comment (trajepl) [reply olive/evaluator/metric.py]: Updated.

File: olive/evaluator/metric_backend.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    @staticmethod
+    def _default_config() -> Dict[str, ConfigParam]:
+        return {
+            "load_params": ConfigParam(
+                type_=Dict[str, Any], default_value=None, description="The parameters to load the metric."
+            ),
+            "compute_params": ConfigParam(
+                type_=Dict[str, Any], default_value=None, description="The parameters to compute the metric."
+            ),
+            "post_lambda_str": ConfigParam(
```
[the remark this answers is not in the document]
Comment (trajepl) [reply olive/evaluator/metric_backend.py]: Updated to remove eval and lambda function. Turn to formatted key to get the value as `evalaute.compute` always return dict.

File: docs/source/overview/options.md [hunk 5]
```diff
@@ -152,11 +152,20 @@ information of the evaluator contains following items:
...
 
         - `name: str` The name of the subtype. Please refer to [AccuracySubtype](accuracy_sub_type) and [LatencySubtype](latency_sub_type)
         for the supported subtypes. For `custom` type, if the result of the evaluation is a dictionary, the name of the subtype should be the key of the dictionary. Otherwise, the name of the subtype could be any unique string user gives.
 
+        - `metric_config` The parameter config used to measure detailed metrics. Please note that when the `backend` is `huggingface_metrics`, you should see the `metric_config` as dictionary of:
+            - `load_params`: The parameters used to load the metric, run as `evaluator = evaluate.load("word_length", **load_params)`.
+            - `compute_params` The parameters used to compute the metric, run as `evaluator.compute(predictions=preds, references=target, **compute_params)`.
+            - `result_key` The key used to extract the metric result with given format. For example, if the metric result is {'accuracy': {'value': 0.9}}, then the post_key_str should be 'accuracy.value'."
```
Comment (guotuofeng) [inline docs/source/overview/options.md:167]: result_key?

**Pull request merged**